File: sdk/python/omni_connector/storage.py
import base64
import logging
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .client import SdkClient

logger = logging.getLogger(__name__)


class ContentStorage:
    """Content storage that delegates to connector-manager via SDK client."""

    def __init__(self, sdk_client: "SdkClient", sync_run_id: str):
        self._client = sdk_client
        self._sync_run_id = sync_run_id

    async def save(
        self,
        content: str | bytes,
        content_type: str = "text/plain",
    ) -> str:
        """Store content and return content_id (ULID)."""
        if isinstance(content, bytes):
            content = content.decode("utf-8")

        return await self._client.store_content(
            self._sync_run_id,
            content,
            content_type,
        )

    async def extract_and_store_content(
        self,
        data: bytes,
        mime_type: str,
        filename: str | None = None,
    ) -> str:
        """Extract text from binary file content and store it, returning content_id.

        The connector manager handles extraction based on MIME type.
        """
        return await self._client.extract_and_store_content(
            self._sync_run_id,
            data,
            mime_type,
            filename,
        )

    async def save_binary(
        self,
        content: bytes,
        content_type: str = "application/octet-stream",
    ) -> str:
        """Store binary content as base64."""
        encoded = base64.b64encode(content).decode("utf-8")
        return await self._client.store_content(
            self._sync_run_id,
            encoded,
            content_type,
        )

File: sdk/typescript/src/storage.ts
import type { SdkClient } from './client.js';

export class ContentStorage {
  private readonly client: SdkClient;
  private readonly syncRunId: string;

  constructor(client: SdkClient, syncRunId: string) {
    this.client = client;
    this.syncRunId = syncRunId;
  }

  async save(content: string, contentType = 'text/plain'): Promise<string> {
    return this.client.storeContent(this.syncRunId, content, contentType);
  }

  async extractAndStoreContent(
    data: Buffer | Uint8Array,
    mimeType: string,
    filename?: string
  ): Promise<string> {
    return this.client.extractAndStoreContent(this.syncRunId, data, mimeType, filename);
  }

  async saveBinary(
    content: Buffer,
    contentType = 'application/octet-stream'
  ): Promise<string> {
    const encoded = content.toString('base64');
    return this.client.storeContent(this.syncRunId, encoded, contentType);
  }
}
